include/seastar/coroutine/async_generator.hh

```diff
@@ -0,0 +1,461 @@
+            }
+
+            iterator await_resume() {
+                if (_promise == nullptr) {
+                    return iterator{nullptr};
+                }
+                if (_promise->finished()) {
+                    _promise->rethrow_if_unhandled_exception();
+                    return iterator{nullptr};
+                }
+                return iterator{
+                    handle_type::from_promise(*static_cast<promise_type *>(_promise))
+                };
+            }
+        };
+
+        if (_coro) {
+            return begin_awaiter{_coro};
+        } else {
+            return begin_awaiter{nullptr};
+        }
+    }
+
+    [[nodiscard]] std::default_sentinel_t end() const noexcept {
+        return {};
+    }
+};
+
+template <typename Ref, typename Value>
+class async_generator<Ref, Value>::promise_type final : public internal::async_generator_promise_base<yielded_type> {
+public:
+    async_generator get_return_object() noexcept {
+        return async_generator{*this};
+    }
+
+    yielded_type value() const noexcept {
+        return static_cast<yielded_type>(*this->_value);
+    }
+};
+
+template <typename Ref, typename Value>
+class async_generator<Ref, Value>::iterator final {
+private:
+    using handle_type = async_generator::handle_type;
+    handle_type _coro = nullptr;
+
+public:
+    using iterator_category = std::input_iterator_tag;
+    using difference_type = std::ptrdiff_t;
+    using value_type = async_generator::value_type;
+    using reference = async_generator::reference_type;
+    using pointer = std::add_pointer_t<value_type>;
+
+    explicit iterator(handle_type coroutine) noexcept
+        : _coro{coroutine}
+    {}
+
+    explicit operator bool() const noexcept {
+        return _coro && !_coro.done();
+    }
+
+    [[nodiscard]] auto operator++() noexcept {
+        using base_awaiter = internal::next_awaiter<yielded_type>;
+        class increment_awaiter final : public base_awaiter {
+            iterator& _iterator;
+            using base_awaiter::_promise;
+
+        public:
+            explicit increment_awaiter(iterator& iterator) noexcept
+                : base_awaiter{iterator._coro.promise(), iterator._coro}
+                , _iterator{iterator}
+            {}
+            iterator& await_resume() {
+                if (_promise->finished()) {
+                    // update iterator to end()
+                    _iterator = iterator{nullptr};
+                    _promise->rethrow_if_unhandled_exception();
+                }
+                return _iterator;
+            }
+        };
+
+        assert(bool(*this) && "cannot increment end iterator");
+        return increment_awaiter{*this};
+    }
+
+    reference operator*() const noexcept {
+        return _coro.promise().value();
+    }
+
+    bool operator==(std::default_sentinel_t) const noexcept {
+        return !bool(*this);
+    }
+};
+
+} // namespace seastar::coroutine::experimental
```

Comment:
Where is std::coroutine_traits specialized?

Reply:
we don't need to specialize it, because the one provided by `<coroutine>` is good enough. according to https://eel.is/c++draft/coroutine.traits.primary

> The header [<coroutine>](https://eel.is/c++draft/coroutine.syn#header:%3ccoroutine%3e) defines the primary template coroutine_traits such that if ArgTypes is a parameter pack of types and if the [qualified-id](https://eel.is/c++draft/expr.prim.id.qual#nt:qualified-id) R​::​promise_type is valid and denotes a type ([[temp.deduct]](https://eel.is/c++draft/temp.deduct)), then coroutine_traits<R, ArgTypes...> has the following publicly accessible member: using promise_type = typename R::promise_type;
> Otherwise, coroutine_traits<R, ArgTypes...> has no members[.](https://eel.is/c++draft/coroutine.traits.primary#1.sentence-2)
> [2](https://eel.is/c++draft/coroutine.traits.primary#2)
> [#](http://github.com/Eelis/draft/tree/0bfdc5feca1e4a216d212dad2ef108eee637a949/source/support.tex#L5219)

since our generator has a public accessible nested type named `promise_type` , the primary template of `coroutine_traits` is able to provide the necessary information -- the `promise_type` of `seastar::coroutine::experimental::generator<...>` , and hence meets the requirements of coroutine.

Reply:
I see. Didn't know about the standard specialization.